File: ios/Classes/NotificationManagerPlugin.swift
import Flutter
import UIKit
import UserNotifications
import Foundation

public class NotificationManagerPlugin: NSObject, FlutterPlugin, UNUserNotificationCenterDelegate {
  private var eventSink: FlutterEventSink?
  private var methodChannel: FlutterMethodChannel?
  private var eventChannel: FlutterEventChannel?
  private let userDefaults = UserDefaults.standard
  private let duplicateKeyPrefix = "notification_duplicate_"
  private let scheduledKeyPrefix = "scheduled_notification_"
  
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "notification_manager", binaryMessenger: registrar.messenger())
    let eventChannel = FlutterEventChannel(name: "notification_manager_events", binaryMessenger: registrar.messenger())
    let instance = NotificationManagerPlugin()
    
    registrar.addMethodCallDelegate(instance, channel: channel)
    eventChannel.setStreamHandler(instance)
    
    instance.methodChannel = channel
    instance.eventChannel = eventChannel
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "initialize":
      initialize(result: result)
    case "requestPermissions":
      requestPermissions(result: result)
    case "areNotificationsEnabled":
      areNotificationsEnabled(result: result)
    case "showNotification":
      showNotification(call: call, result: result)
    case "scheduleNotification":
      scheduleNotification(call: call, result: result)
    case "getScheduledNotifications":
      getScheduledNotifications(result: result)
    case "updateScheduledNotification":
      updateScheduledNotification(call: call, result: result)
    case "cancelNotification":
      cancelNotification(call: call, result: result)
    case "cancelScheduledNotification":
      cancelScheduledNotification(call: call, result: result)
    case "cancelAllNotifications":
      cancelAllNotifications(result: result)
    case "cancelAllScheduledNotifications":
      cancelAllScheduledNotifications(result: result)
    case "getBadgeCount":
      getBadgeCount(result: result)
    case "setBadgeCount":
      setBadgeCount(call: call, result: result)
    case "clearBadgeCount":
      clearBadgeCount(result: result)
    case "isDuplicateNotification":
      isDuplicateNotification(call: call, result: result)
    case "clearNotificationHistory":
      clearNotificationHistory(result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }
  
  private func initialize(result: @escaping FlutterResult) {
    UNUserNotificationCenter.current().delegate = self
    result(true)
  }
  
  private func requestPermissions(result: @escaping FlutterResult) {
    UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .badge, .sound]) { granted, error in
      DispatchQueue.main.async {
        result(granted)
      }
    }
  }
  
  private func areNotificationsEnabled(result: @escaping FlutterResult) {
    UNUserNotificationCenter.current().getNotificationSettings { settings in
      DispatchQueue.main.async {
        result(settings.authorizationStatus == .authorized)
      }
    }
  }
  
  private func showNotification(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let arguments = call.arguments as? [String: Any],
          let id = arguments["id"] as? String,
          let title = arguments["title"] as? String,
          let body = arguments["body"] as? String else {
      result(false)
      return
    }
    
    let actions = arguments["actions"] as? [[String: Any]] ?? []
    let payload = arguments["payload"] as? [String: Any]
    let category = arguments["category"] as? String ?? "default"
    let badgeNumber = arguments["badgeNumber"] as? Int
    let duplicateKey = arguments["duplicateKey"] as? String
    let duplicateWindow = arguments["duplicateWindow"] as? Int
    
    // Check for duplicates if duplicateKey is provided
    if let duplicateKey = duplicateKey {
      if isDuplicateNotificationInternal(duplicateKey: duplicateKey, timeWindow: duplicateWindow) {
        result(false)
        return
      }
      markNotificationAsSent(duplicateKey: duplicateKey)
    }
    
    // Create notification content
    let content = UNMutableNotificationContent()
    content.title = title
    content.body = body
    content.sound = .default
    
    if let payload = payload {
      content.userInfo = payload
    }
    
    if let badgeNumber = badgeNumber {
      content.badge = NSNumber(value: badgeNumber)
    }
    
    // Create notification category with actions
    if !actions.isEmpty {
      var notificationActions: [UNNotificationAction] = []
      
      for actionData in actions {
        guard let actionId = actionData["id"] as? String,
              let actionTitle = actionData["title"] as? String else { continue }
        
        let isDestructive = actionData["isDestructive"] as? Bool ?? false
        let requiresAuthentication = actionData["requiresAuthentication"] as? Bool ?? false
        
        let action = UNNotificationAction(
          identifier: actionId,
          title: actionTitle,
          options: isDestructive ? [.destructive] : []
        )
        notificationActions.append(action)
      }
      
      let notificationCategory = UNNotificationCategory(
        identifier: category,
        actions: notificationActions,
        intentIdentifiers: [],
        options: []
      )
      
      UNUserNotificationCenter.current().setNotificationCategories([notificationCategory])
      content.categoryIdentifier = notificationCategory.identifier
    }
    
    // Create trigger (immediate)
    let trigger = UNTimeIntervalNotificationTrigger(timeInterval: 1, repeats: false)
    
    // Create request
    let request = UNNotificationRequest(
      identifier: id,
      content: content,
      trigger: trigger
    )
    
    // Schedule notification
    UNUserNotificationCenter.current().add(request) { error in
      DispatchQueue.main.async {
        if error != nil {
          result(false)
        } else {
          result(true)
        }
      }
    }
  }
  
  private func scheduleNotification(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let arguments = call.arguments as? [String: Any],
          let id = arguments["id"] as? String,
          let requestData = arguments["request"] as? [String: Any],
          let title = requestData["title"] as? String,
          let body = requestData["body"] as? String,
          let scheduledDate = arguments["scheduledDate"] as? Double else {
      result(false)
      return
    }
    
    let actions = requestData["actions"] as? [[String: Any]] ?? []
    let payload = requestData["payload"] as? [String: Any]
    let category = requestData["category"] as? String ?? "default"
    let badgeNumber = requestData["badgeNumber"] as? Int
    let isRepeating = arguments["isRepeating"] as? Bool ?? false
    let repeatInterval = arguments["repeatInterval"] as? Int
    
    // Calculate time interval
    let scheduledTime = Date(timeIntervalSince1970: scheduledDate / 1000.0)
    let timeInterval = scheduledTime.timeIntervalSinceNow
    
    if timeInterval <= 0 {
      result(false)
      return
    }
    
    // Create notification content
    let content = UNMutableNotificationContent()
    content.title = title
    content.body = body
    content.sound = .default
    
    if let payload = payload {
      content.userInfo = payload
    }
    
    if let badgeNumber = badgeNumber {
      content.badge = NSNumber(value: badgeNumber)
    }
    
    // Create notification category with actions
    if !actions.isEmpty {
      var notificationActions: [UNNotificationAction] = []
      
      for actionData in actions {
        guard let actionId = actionData["id"] as? String,
              let actionTitle = actionData["title"] as? String else { continue }
        
        let isDestructive = actionData["isDestructive"] as? Bool ?? false
        
        let action = UNNotificationAction(
          identifier: actionId,
          title: actionTitle,
          options: isDestructive ? [.destructive] : []
        )
        notificationActions.append(action)
      }
      
      let notificationCategory = UNNotificationCategory(
        identifier: category,
        actions: notificationActions,
        intentIdentifiers: [],
        options: []
      )
      
      UNUserNotificationCenter.current().setNotificationCategories([notificationCategory])
      content.categoryIdentifier = notificationCategory.identifier
    }
    
    // Create trigger
    let trigger: UNNotificationTrigger
    if isRepeating, let repeatInterval = repeatInterval {
      trigger = UNTimeIntervalNotificationTrigger(timeInterval: TimeInterval(repeatInterval), repeats: true)
    } else {
      trigger = UNTimeIntervalNotificationTrigger(timeInterval: timeInterval, repeats: false)
    }
    
    // Create request
    let request = UNNotificationRequest(
      identifier: id,
      content: content,
      trigger: trigger
    )
    
    // Schedule notification
    UNUserNotificationCenter.current().add(request) { error in
      DispatchQueue.main.async {
        if error != nil {
          result(false)
        } else {
          // Store scheduled notification info
          let scheduledNotification: [String: Any] = [
            "id": id,
            "scheduledDate": scheduledDate,
            "isRepeating": isRepeating,
            "repeatInterval": repeatInterval ?? 0,
            "request": requestData
          ]
          
          if let data = try? JSONSerialization.data(withJSONObject: scheduledNotification),
             let jsonString = String(data: data, encoding: .utf8) {
            self.userDefaults.set(jsonString, forKey: "\(self.scheduledKeyPrefix)\(id)")
          }
          
          result(true)
        }
      }
    }
  }
  
  private func getScheduledNotifications(result: @escaping FlutterResult) {
    UNUserNotificationCenter.current().getPendingNotificationRequests { requests in
      var scheduledNotifications: [[String: Any]] = []
      
      for request in requests {
        if let scheduledData = self.userDefaults.string(forKey: "\(self.scheduledKeyPrefix)\(request.identifier)"),
           let data = scheduledData.data(using: .utf8),
           let notificationData = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
          scheduledNotifications.append(notificationData)
        }
      }
      
      DispatchQueue.main.async {
        result(scheduledNotifications)
      }
    }
  }
  
  private func updateScheduledNotification(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let arguments = call.arguments as? [String: Any],
          let id = arguments["id"] as? String else {
      result(false)
      return
    }
    
    // Cancel existing notification
    UNUserNotificationCenter.current().removePendingNotificationRequests(withIdentifiers: [id])
    
    // Schedule new notification with updated data
    scheduleNotification(call: call, result: result)
  }
  
  private func cancelNotification(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let arguments = call.arguments as? [String: Any],
          let id = arguments["id"] as? String else {
      result(false)
      return
    }
    
    UNUserNotificationCenter.current().removePendingNotificationRequests(withIdentifiers: [id])
    UNUserNotificationCenter.current().removeDeliveredNotifications(withIdentifiers: [id])
    result(true)
  }
  
  private func cancelScheduledNotification(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let arguments = call.arguments as? [String: Any],
          let id = arguments["id"] as? String else {
      result(false)
      return
    }
    
    UNUserNotificationCenter.current().removePendingNotificationRequests(withIdentifiers: [id])
    userDefaults.removeObject(forKey: "\(scheduledKeyPrefix)\(id)")
    result(true)
  }
  
  private func cancelAllNotifications(result: @escaping FlutterResult) {
    UNUserNotificationCenter.current().removeAllPendingNotificationRequests()
    UNUserNotificationCenter.current().removeAllDeliveredNotifications()
    result(true)
  }
  
  private func cancelAllScheduledNotifications(result: @escaping FlutterResult) {
    UNUserNotificationCenter.current().removeAllPendingNotificationRequests()
    
    // Clear all scheduled notification data
    let allKeys = userDefaults.dictionaryRepresentation().keys
    for key in allKeys {
      if key.hasPrefix(scheduledKeyPrefix) {
        userDefaults.removeObject(forKey: key)
      }
    }
    
    result(true)
  }
  
  private func getBadgeCount(result: @escaping FlutterResult) {
    result(UIApplication.shared.applicationIconBadgeNumber)
  }
  
  private func setBadgeCount(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let arguments = call.arguments as? [String: Any],
          let count = arguments["count"] as? Int else {
      result(false)
      return
    }
    
    UIApplication.shared.applicationIconBadgeNumber = count
    result(true)
  }
  
  private func clearBadgeCount(result: @escaping FlutterResult) {
    UIApplication.shared.applicationIconBadgeNumber = 0
    result(true)
  }
  
  private func isDuplicateNotification(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let arguments = call.arguments as? [String: Any],
          let duplicateKey = arguments["duplicateKey"] as? String else {
      result(false)
      return
    }
    
    let timeWindow = arguments["timeWindow"] as? Int
    let isDuplicate = isDuplicateNotificationInternal(duplicateKey: duplicateKey, timeWindow: timeWindow)
    result(isDuplicate)
  }
  
  private func clearNotificationHistory(result: @escaping FlutterResult) {
    UNUserNotificationCenter.current().removeAllDeliveredNotifications()
    
    // Clear all duplicate tracking data
    let allKeys = userDefaults.dictionaryRepresentation().keys
    for key in allKeys {
      if key.hasPrefix(duplicateKeyPrefix) {
        userDefaults.removeObject(forKey: key)
      }
    }
    
    result(true)
  }
  
  // MARK: - Helper Methods
  
  private func isDuplicateNotificationInternal(duplicateKey: String, timeWindow: Int?) -> Bool {
    let key = "\(duplicateKeyPrefix)\(duplicateKey)"
    let lastSentTime = userDefaults.double(forKey: key)
    
    if lastSentTime == 0 {
      return false
    }
    
    let currentTime = Date().timeIntervalSince1970
    let window = timeWindow ?? 300 // Default 5 minutes
    
    return (currentTime - lastSentTime) < Double(window)
  }
  
  private func markNotificationAsSent(duplicateKey: String) {
    let key = "\(duplicateKeyPrefix)\(duplicateKey)"
    userDefaults.set(Date().timeIntervalSince1970, forKey: key)
  }
  
  // MARK: - UNUserNotificationCenterDelegate
  
  public func userNotificationCenter(_ center: UNUserNotificationCenter, willPresent notification: UNNotification, withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
    completionHandler([.alert, .badge, .sound])
  }
  
  public func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
    let notificationId = response.notification.request.identifier
    let userInfo = response.notification.request.content.userInfo
    
    if response.actionIdentifier == UNNotificationDefaultActionIdentifier {
      // Notification was tapped
      let event: [String: Any] = [
        "type": "tap",
        "notificationId": notificationId,
        "payload": userInfo
      ]
      eventSink?(event)
    } else {
      // Action button was tapped
      let event: [String: Any] = [
        "type": "action",
        "notificationId": notificationId,
        "actionId": response.actionIdentifier
      ]
      eventSink?(event)
    }
    
    completionHandler()
  }
}

// MARK: - FlutterStreamHandler

extension NotificationManagerPlugin: FlutterStreamHandler {
  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    self.eventSink = events
    return nil
  }
  
  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    self.eventSink = nil
    return nil
  }
}

File: android/src/main/kotlin/com/example/notification_manager/NotificationManagerPlugin.kt
package com.example.notification_manager

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.SharedPreferences
import android.os.Build
import androidx.core.app.NotificationCompat
import androidx.work.Data
import androidx.work.OneTimeWorkRequestBuilder
import androidx.work.WorkManager
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/** NotificationManagerPlugin */
class NotificationManagerPlugin: FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  private lateinit var channel: MethodChannel
  private lateinit var eventChannel: EventChannel
  private lateinit var context: Context
  private lateinit var notificationManager: NotificationManager
  private lateinit var sharedPreferences: SharedPreferences
  private var eventSink: EventChannel.EventSink? = null

  companion object {
    private const val CHANNEL_ID = "notification_manager_channel"
    private const val CHANNEL_NAME = "Notification Manager"
    private const val CHANNEL_DESCRIPTION = "Notifications from Flutter app"
    private const val ACTION_NOTIFICATION_TAP = "notification_tap"
    private const val ACTION_NOTIFICATION_ACTION = "notification_action"
    private const val PREF_NAME = "notification_manager_prefs"
    private const val PREF_DUPLICATE_PREFIX = "duplicate_"
  }

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    context = flutterPluginBinding.applicationContext
    notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
    sharedPreferences = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
    
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "notification_manager")
    channel.setMethodCallHandler(this)
    
    eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "notification_manager_events")
    eventChannel.setStreamHandler(this)
    
    createNotificationChannel()
    registerBroadcastReceivers()
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "initialize" -> {
        result.success(true)
      }
      "requestPermissions" -> {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
          // For Android 13+, we need to request POST_NOTIFICATIONS permission
          val permission = android.Manifest.permission.POST_NOTIFICATIONS
          if (context.checkSelfPermission(permission) != android.content.pm.PackageManager.PERMISSION_GRANTED) {
            // Permission not granted, return false
            result.success(false)
          } else {
            result.success(true)
          }
        } else {
          // For older versions, notifications are enabled by default
          result.success(true)
        }
      }
      "areNotificationsEnabled" -> {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
          val permission = android.Manifest.permission.POST_NOTIFICATIONS
          val hasPermission = context.checkSelfPermission(permission) == android.content.pm.PackageManager.PERMISSION_GRANTED
          result.success(hasPermission)
        } else {
          // For older versions, assume notifications are enabled
          result.success(true)
        }
      }
      "showNotification" -> {
        val arguments = call.arguments as Map<*, *>
        showNotification(arguments, result)
      }
      "scheduleNotification" -> {
        val arguments = call.arguments as Map<*, *>
        scheduleNotification(arguments, result)
      }
      "getScheduledNotifications" -> {
        getScheduledNotifications(result)
      }
      "updateScheduledNotification" -> {
        val arguments = call.arguments as Map<*, *>
        updateScheduledNotification(arguments, result)
      }
      "cancelNotification" -> {
        val id = call.argument<String>("id")
        if (id != null) {
          notificationManager.cancel(id.hashCode())
          result.success(true)
        } else {
          result.success(false)
        }
      }
      "cancelScheduledNotification" -> {
        val id = call.argument<String>("id")
        if (id != null) {
          cancelScheduledNotification(id, result)
        } else {
          result.success(false)
        }
      }
      "cancelAllNotifications" -> {
        notificationManager.cancelAll()
        result.success(true)
      }
      "cancelAllScheduledNotifications" -> {
        cancelAllScheduledNotifications(result)
      }
      "getBadgeCount" -> {
        // Android doesn't have a built-in badge count, return 0
        result.success(0)
      }
      "setBadgeCount" -> {
        val count = call.argument<Int>("count") ?: 0
        // Android doesn't have a built-in badge count, just return success
        result.success(true)
      }
      "clearBadgeCount" -> {
        // Android doesn't have a built-in badge count, just return success
        result.success(true)
      }
      "isDuplicateNotification" -> {
        val duplicateKey = call.argument<String>("duplicateKey")
        val timeWindow = call.argument<Int>("timeWindow")
        if (duplicateKey != null) {
          val isDuplicate = checkDuplicateNotification(duplicateKey, timeWindow)
          result.success(isDuplicate)
        } else {
          result.success(false)
        }
      }
      "clearNotificationHistory" -> {
        clearNotificationHistory(result)
      }
      else -> {
        result.notImplemented()
      }
    }
  }

  private fun createNotificationChannel() {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      val channel = NotificationChannel(
        CHANNEL_ID,
        CHANNEL_NAME,
        NotificationManager.IMPORTANCE_HIGH
      ).apply {
        description = CHANNEL_DESCRIPTION
        enableLights(true)
        enableVibration(true)
        setShowBadge(true)
        lockscreenVisibility = Notification.VISIBILITY_PUBLIC
        setSound(android.provider.Settings.System.DEFAULT_NOTIFICATION_URI, null)
      }
      notificationManager.createNotificationChannel(channel)
      android.util.Log.d("NotificationManager", "Notification channel created: $CHANNEL_ID with importance: ${channel.importance}")
    }
  }

  private fun registerBroadcastReceivers() {
    val tapFilter = IntentFilter(ACTION_NOTIFICATION_TAP)
    val actionFilter = IntentFilter(ACTION_NOTIFICATION_ACTION)
    
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
      context.registerReceiver(notificationTapReceiver, tapFilter, Context.RECEIVER_NOT_EXPORTED)
      context.registerReceiver(notificationActionReceiver, actionFilter, Context.RECEIVER_NOT_EXPORTED)
    } else {
      context.registerReceiver(notificationTapReceiver, tapFilter)
      context.registerReceiver(notificationActionReceiver, actionFilter)
    }
  }

  private fun showNotification(arguments: Map<*, *>, result: Result) {
    try {
      // Debug: Log the arguments received
      android.util.Log.d("NotificationManager", "Received arguments: $arguments")
      
      // Extract fields from NotificationRequest structure
      val id = arguments["id"] as String
      val title = arguments["title"] as String
      val body = arguments["body"] as String
      val actions = arguments["actions"] as? List<Map<*, *>>
      val payload = arguments["payload"] as? Map<*, *>
      val badgeNumber = arguments["badgeNumber"] as? Int
      val duplicateKey = arguments["duplicateKey"] as? String
      val duplicateWindow = arguments["duplicateWindow"] as? Int

      // Check for duplicates if duplicateKey is provided
      if (duplicateKey != null) {
        val isDuplicate = checkDuplicateNotification(duplicateKey, duplicateWindow)
        if (isDuplicate) {
          result.success(false)
          return
        }
        // Mark this notification as sent
        markNotificationAsSent(duplicateKey)
      }

      // Ensure notification channel exists
      createNotificationChannel()

      val builder = NotificationCompat.Builder(context, CHANNEL_ID)
        .setContentTitle(title)
        .setContentText(body)
        .setSmallIcon(android.R.drawable.ic_dialog_info)
        .setPriority(NotificationCompat.PRIORITY_HIGH)
        .setAutoCancel(true)
        .setDefaults(NotificationCompat.DEFAULT_ALL)

      // Add tap intent
      val tapIntent = Intent(ACTION_NOTIFICATION_TAP).apply {
        putExtra("notification_id", id)
        if (payload != null) {
          putExtra("payload", JSONObject(payload as Map<String, Any>).toString())
        }
      }
      val tapPendingIntent = PendingIntent.getBroadcast(
        context,
        id.hashCode(),
        tapIntent,
        PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
      )
      builder.setContentIntent(tapPendingIntent)

      // Add action buttons
      actions?.forEachIndexed { index, action ->
        val actionId = action["id"] as String
        val actionTitle = action["title"] as String
        val isDestructive = action["isDestructive"] as? Boolean ?: false

        val actionIntent = Intent(ACTION_NOTIFICATION_ACTION).apply {
          putExtra("notification_id", id)
          putExtra("action_id", actionId)
        }
        val actionPendingIntent = PendingIntent.getBroadcast(
          context,
          (id + actionId).hashCode(),
          actionIntent,
          PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )

        builder.addAction(
          if (isDestructive) android.R.drawable.ic_menu_delete else android.R.drawable.ic_menu_send,
          actionTitle,
          actionPendingIntent
        )
      }

      // Set badge number if provided
      if (badgeNumber != null && badgeNumber > 0) {
        builder.setNumber(badgeNumber)
      }

      val notification = builder.build()
      notificationManager.notify(id.hashCode(), notification)
      
      // Log for debugging
      android.util.Log.d("NotificationManager", "Notification sent: $id - $title")
      
      result.success(true)
    } catch (e: Exception) {
      android.util.Log.e("NotificationManager", "Error showing notification: ${e.message}")
      result.error("SHOW_NOTIFICATION_ERROR", e.message, null)
    }
  }

  private fun scheduleNotification(arguments: Map<*, *>, result: Result) {
    try {
      val id = arguments["id"] as String
      val request = arguments["request"] as Map<*, *>
      val scheduledDate = arguments["scheduledDate"] as Long
      val isRepeating = arguments["isRepeating"] as? Boolean ?: false
      val repeatInterval = arguments["repeatInterval"] as? Int

      val delayMillis = scheduledDate - System.currentTimeMillis()
      if (delayMillis <= 0) {
        result.success(false)
        return
      }

      // Create work data
      val workData = Data.Builder()
        .putString("id", id)
        .putString("title", request["title"] as String)
        .putString("body", request["body"] as String)
        .putString("actions", request["actions"]?.toString() ?: "[]")
        .putString("payload", request["payload"]?.toString() ?: "{}")
        .putInt("badgeNumber", request["badgeNumber"] as? Int ?: 0)
        .putBoolean("isRepeating", isRepeating)
        .putInt("repeatInterval", repeatInterval ?: 0)
        .build()

      // Create work request
      val workRequest = OneTimeWorkRequestBuilder<NotificationWorker>()
        .setInputData(workData)
        .setInitialDelay(delayMillis, TimeUnit.MILLISECONDS)
        .addTag("notification_$id")
        .build()

      // Schedule the work
      WorkManager.getInstance(context).enqueue(workRequest)

      // Store scheduled notification info
      val scheduledNotification = mapOf(
        "id" to id,
        "scheduledDate" to scheduledDate,
        "isRepeating" to isRepeating,
        "repeatInterval" to repeatInterval,
        "request" to request
      )
      sharedPreferences.edit().putString("scheduled_$id", JSONObject(scheduledNotification).toString()).apply()

      result.success(true)
    } catch (e: Exception) {
      result.error("SCHEDULE_NOTIFICATION_ERROR", e.message, null)
    }
  }

  private fun getScheduledNotifications(result: Result) {
    try {
      val scheduledNotifications = mutableListOf<Map<String, Any>>()
      val allPrefs = sharedPreferences.all

      for ((key, value) in allPrefs) {
        if (key.startsWith("scheduled_")) {
          val notificationData = JSONObject(value.toString())
          scheduledNotifications.add(notificationData.toMap())
        }
      }

      result.success(scheduledNotifications)
    } catch (e: Exception) {
      result.error("GET_SCHEDULED_NOTIFICATIONS_ERROR", e.message, null)
    }
  }

  private fun updateScheduledNotification(arguments: Map<*, *>, result: Result) {
    try {
      val id = arguments["id"] as String
      val scheduledDate = arguments["scheduledDate"] as Long
      val isRepeating = arguments["isRepeating"] as? Boolean ?: false
      val repeatInterval = arguments["repeatInterval"] as? Int

      // Cancel existing scheduled notification
      WorkManager.getInstance(context).cancelAllWorkByTag("notification_$id")

      // Schedule updated notification
      val delayMillis = scheduledDate - System.currentTimeMillis()
      if (delayMillis > 0) {
        val workData = Data.Builder()
          .putString("id", id)
          .putLong("scheduledDate", scheduledDate)
          .putBoolean("isRepeating", isRepeating)
          .putInt("repeatInterval", repeatInterval ?: 0)
          .build()

        val workRequest = OneTimeWorkRequestBuilder<NotificationWorker>()
          .setInputData(workData)
          .setInitialDelay(delayMillis, TimeUnit.MILLISECONDS)
          .addTag("notification_$id")
          .build()

        WorkManager.getInstance(context).enqueue(workRequest)

        // Update stored notification info
        val updatedNotification = mapOf(
          "id" to id,
          "scheduledDate" to scheduledDate,
          "isRepeating" to isRepeating,
          "repeatInterval" to repeatInterval
        )
        sharedPreferences.edit().putString("scheduled_$id", JSONObject(updatedNotification).toString()).apply()
      }

      result.success(true)
    } catch (e: Exception) {
      result.error("UPDATE_SCHEDULED_NOTIFICATION_ERROR", e.message, null)
    }
  }

  private fun cancelScheduledNotification(id: String, result: Result) {
    try {
      // Cancel work
      WorkManager.getInstance(context).cancelAllWorkByTag("notification_$id")
      
      // Remove from storage
      sharedPreferences.edit().remove("scheduled_$id").apply()
      
      result.success(true)
    } catch (e: Exception) {
      result.error("CANCEL_SCHEDULED_NOTIFICATION_ERROR", e.message, null)
    }
  }

  private fun cancelAllScheduledNotifications(result: Result) {
    try {
      // Cancel all scheduled work
      WorkManager.getInstance(context).cancelAllWork()
      
      // Clear all scheduled notifications from storage
      val allPrefs = sharedPreferences.all
      val editor = sharedPreferences.edit()
      for ((key, _) in allPrefs) {
        if (key.startsWith("scheduled_")) {
          editor.remove(key)
        }
      }
      editor.apply()
      
      result.success(true)
    } catch (e: Exception) {
      result.error("CANCEL_ALL_SCHEDULED_NOTIFICATIONS_ERROR", e.message, null)
    }
  }

  private fun checkDuplicateNotification(duplicateKey: String, timeWindowSeconds: Int?): Boolean {
    val key = PREF_DUPLICATE_PREFIX + duplicateKey
    val lastSentTime = sharedPreferences.getLong(key, 0)
    val currentTime = System.currentTimeMillis()
    val windowMillis = (timeWindowSeconds ?: 300) * 1000L // Default 5 minutes

    return (currentTime - lastSentTime) < windowMillis
  }

  private fun markNotificationAsSent(duplicateKey: String) {
    val key = PREF_DUPLICATE_PREFIX + duplicateKey
    sharedPreferences.edit().putLong(key, System.currentTimeMillis()).apply()
  }

  private fun clearNotificationHistory(result: Result) {
    try {
      // Clear all notifications
      notificationManager.cancelAll()
      
      // Clear duplicate tracking
      val allPrefs = sharedPreferences.all
      val editor = sharedPreferences.edit()
      for ((key, _) in allPrefs) {
        if (key.startsWith(PREF_DUPLICATE_PREFIX)) {
          editor.remove(key)
        }
      }
      editor.apply()
      
      result.success(true)
    } catch (e: Exception) {
      result.error("CLEAR_NOTIFICATION_HISTORY_ERROR", e.message, null)
    }
  }

  private val notificationTapReceiver = object : BroadcastReceiver() {
    override fun onReceive(context: Context?, intent: Intent?) {
      val notificationId = intent?.getStringExtra("notification_id")
      val payloadString = intent?.getStringExtra("payload")
      
      val event = mapOf(
        "type" to "tap",
        "notificationId" to notificationId,
        "payload" to if (payloadString != null) JSONObject(payloadString).toMap() else null
      )
      
      eventSink?.success(event)
    }
  }

  private val notificationActionReceiver = object : BroadcastReceiver() {
    override fun onReceive(context: Context?, intent: Intent?) {
      val notificationId = intent?.getStringExtra("notification_id")
      val actionId = intent?.getStringExtra("action_id")
      
      val event = mapOf(
        "type" to "action",
        "notificationId" to notificationId,
        "actionId" to actionId
      )
      
      eventSink?.success(event)
    }
  }

  override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
    eventSink = events
  }

  override fun onCancel(arguments: Any?) {
    eventSink = null
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
    eventChannel.setStreamHandler(null)
    
    try {
      context.unregisterReceiver(notificationTapReceiver)
      context.unregisterReceiver(notificationActionReceiver)
    } catch (e: Exception) {
      // Receiver might not be registered
    }
  }

  private fun JSONObject.toMap(): Map<String, Any> {
    val map = mutableMapOf<String, Any>()
    val keys = this.keys()
    while (keys.hasNext()) {
      val key = keys.next()
      map[key] = this.get(key)
    }
    return map
  }
}
